edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java

```diff
@@ -120,44 +120,52 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
             return false;
         }
 
-
         var bpn = participantAgent.getIdentity();
         var groups = store.resolveForBpn(bpn);
 
         // BPN not found in database
         if (groups.failed()) {
-            policyContext.reportProblem(groups.getFailureDetail());
-            return false;
+            return evaluateEmptyGroupsResponse(operator, policyContext, groups.getFailureDetail());
         }
+
         var assignedGroups = groups.getContent();
 
         // BPN was found, but it does not have groups assigned.
         if (assignedGroups.isEmpty()) {
-            policyContext.reportProblem("No groups were assigned to BPN " + bpn);
-            return false;
+            return evaluateEmptyGroupsResponse(operator, policyContext, "No groups were assigned to BPN " + bpn);
         }
 
         // right-operand is anything other than String or Collection
         var rightOperand = parseRightOperand(rightValue, policyContext);
-        if (rightOperand == null) {
+        if (rightOperand.isEmpty()) {
             return false;
         }
 
         //call evaluator function
         return OPERATOR_EVALUATOR_MAP.get(operator).apply(new BpnGroupHolder(assignedGroups, rightOperand));
     }
 
+    private boolean evaluateEmptyGroupsResponse(final Operator operator, final PolicyContext policyContext,
+                                                final String problemMessage) {
+        if (NEQ.equals(operator) || IS_NONE_OF.equals(operator)) {
+            return true;
+        }
+
+        policyContext.reportProblem(problemMessage);
+        return false;
+    }
+
     private List<String> parseRightOperand(Object rightValue, PolicyContext context) {
-        if (rightValue instanceof String) {
-            var tokens = ((String) rightValue).split(",");
+        if (rightValue instanceof String value) {
+            var tokens = value.split(",");
             return Arrays.asList(tokens);
         }
         if (rightValue instanceof Collection<?>) {
             return ((Collection<?>) rightValue).stream().map(Object::toString).toList();
         }
 
-        context.reportProblem(format("Right operand expected to be either String or a Collection, but was " + rightValue.getClass()));
-        return null;
+        context.reportProblem(format("Right operand expected to be either String or a Collection, but was %s", rightValue.getClass()));
+        return Collections.emptyList();
```

Comment:
is there a reason for this? how can you then distinguish between an empty array `"[]"` and an invalid right operand?

Reply:
It was more of a suggestion from Sonar to return an empty list instead of a null value, then I changed the validation after the method to verify if the list is empty.